base/linking.jl

```diff
@@ -0,0 +1,174 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+module Linking
+
+import Base.Libc: Libdl
+
+# inlined LLD_jll
+# These get calculated in __init__()
+const PATH = Ref("")
+const LIBPATH = Ref("")
+const PATH_list = String[]
+const LIBPATH_list = String[]
+const lld_path = Ref{String}()
+const lld_exe = Sys.iswindows() ? "lld.exe" : "lld"
+
+if Sys.iswindows()
+    const LIBPATH_env = "PATH"
+    const LIBPATH_default = ""
+    const pathsep = ';'
+elseif Sys.isapple()
+    const LIBPATH_env = "DYLD_FALLBACK_LIBRARY_PATH"
+    const LIBPATH_default = "~/lib:/usr/local/lib:/lib:/usr/lib"
+    const pathsep = ':'
+else
+    const LIBPATH_env = "LD_LIBRARY_PATH"
+    const LIBPATH_default = ""
+    const pathsep = ':'
+end
+
+function adjust_ENV!(env::Dict, PATH::String, LIBPATH::String, adjust_PATH::Bool, adjust_LIBPATH::Bool)
+    if adjust_LIBPATH
+        LIBPATH_base = get(env, LIBPATH_env, expanduser(LIBPATH_default))
+        if !isempty(LIBPATH_base)
+            env[LIBPATH_env] = string(LIBPATH, pathsep, LIBPATH_base)
+        else
+            env[LIBPATH_env] = LIBPATH
+        end
+    end
+    if adjust_PATH && (LIBPATH_env != "PATH" || !adjust_LIBPATH)
+        if adjust_PATH
+            if !isempty(get(env, "PATH", ""))
+                env["PATH"] = string(PATH, pathsep, env["PATH"])
+            else
+                env["PATH"] = PATH
+            end
+        end
+    end
+    return env
+end
+
+function __init_lld_path()
+    # Prefer our own bundled lld, but if we don't have one, pick it up off of the PATH
+    # If this is an in-tree build, `lld` will live in `tools`.  Otherwise, it'll be in `libexec`
+    for bundled_lld_path in (joinpath(Sys.BINDIR, Base.LIBEXECDIR, lld_exe),
+                             joinpath(Sys.BINDIR, "..", "tools", lld_exe),
+                             joinpath(Sys.BINDIR, lld_exe))
+        if isfile(bundled_lld_path)
+            lld_path[] = abspath(bundled_lld_path)
+            return
+        end
+    end
+    lld_path[] = something(Sys.which(lld_exe), lld_exe)
+    return
+end
+
+const AVAILABLE = Ref{Bool}(true)
+function available()
+    return AVAILABLE[]
+end
+
+import Base: DevNull
+function __init__()
+    # utils_available = @static if Sys.isapple()
+    #     true
+    #     # success(pipeline(`xcode-select -p`, DevNull(), DevNull(), DevNull()))
+    # else
+    #     true
+    # end
+    disable = parse(Bool, get(ENV, "JULIA_DISABLE_OBJCACHE", "false"))
+    AVAILABLE[] = !disable # && utils_available
```

Comment:
Note to self, move this out of Linking into Loading, now that it is no longer dependent on xcode utils being available.